File: frontend/src/api/screening.ts
import { ApiClient } from './request'

export interface ScreeningOrderBy { field: string; direction: 'asc' | 'desc' }
export interface ScreeningRunReq {
  market?: 'CN'
  date?: string | null
  adj?: 'qfq' | 'hfq' | 'none'
  conditions: any
  order_by?: ScreeningOrderBy[]
  limit?: number
  offset?: number
}

export interface ScreeningRunItem {
  code: string
  close?: number
  pct_chg?: number
  amount?: number
  ma20?: number
  rsi14?: number
  kdj_k?: number
  kdj_d?: number
  kdj_j?: number
  dif?: number
  dea?: number
  macd_hist?: number
}

export interface ScreeningRunResp { total: number; items: ScreeningRunItem[] }

// 筛选字段配置
export interface FieldInfo {
  name: string
  display_name: string
  field_type: string
  data_type: string
  description: string
  supported_operators: string[]
}

export interface FieldConfigResponse {
  fields: Record<string, FieldInfo>
  categories: Record<string, string[]>
}

// 行业列表响应
export interface IndustryOption {
  value: string
  label: string
  count: number
}

export interface IndustriesResponse {
  industries: IndustryOption[]
  total: number
}

export interface LimitUpPullbackScanReq {
  min_score?: number
  top_n?: number
  hold_days?: number
  initial_capital?: number
  limit?: number
}

export interface LimitUpPullbackBacktestReq extends LimitUpPullbackScanReq {
  start_date?: string
  end_date?: string
}

export interface LimitUpPullbackItem {
  code: string
  name: string
  close: number
  pct_chg: number
  limit_up_date: string
  days_since_limit_up: number
  limit_up_close: number
  pullback_depth: number
  volume_shrink_ratio: number
  ground_volume_ratio: number
  lower_shadow_pct: number
  signal_type: string
  score: number
  score_details: string[]
  upside_space: number
  ma5?: number
  ma10?: number
  ma20?: number
  small_body_ratio: number
  ground_day_offset: number
  industry: string
  shadow_to_body_ratio?: number
  space_position_ok?: boolean
  is_20day_ground?: boolean
  fake_breakout?: boolean
  days_since_ground_day?: number
  rebound_from_ground?: number
  // 各阶段日期
  pullback_start_date?: string
  pullback_end_date?: string
  bottom_watch_start_date?: string
  left_buy_date?: string
  right_buy_date?: string
  ground_day_date?: string
}

export interface LimitUpPullbackScanResp {
  total: number
  items: LimitUpPullbackItem[]
  took_ms?: number
  scanned_count?: number
  params?: Record<string, any>
}

export interface BacktestTrade {
  code: string
  name: string
  buy_date: string
  sell_date: string
  buy_price: number
  sell_price: number
  return_pct: number
  score: number
  signal_type: string
  sell_reason: string
  // 各阶段日期
  limit_up_date?: string
  pullback_start_date?: string
  pullback_end_date?: string
  bottom_watch_start_date?: string
  left_buy_date?: string
  right_buy_date?: string
  ground_day_date?: string
  // 卖点日期
  stop_loss_date?: string
  time_stop_date?: string
  high_stop_date?: string
  ma5_stop_date?: string
}

export interface LimitUpPullbackBacktestResp {
  total_trades: number
  win_rate: number
  avg_return: number
  avg_win: number
  avg_loss: number
  profit_loss_ratio: number
  max_drawdown: number
  sharpe_ratio: number
  calmar_ratio: number
  annualized_return: number
  max_consecutive_losses: number
  total_fees_est: number
  total_return: number
  final_capital: number
  initial_capital: number
  backtest_days: number
  signal_stats: Record<string, { count: number; win_rate: number; avg_return: number }>
  sell_reason_stats: Record<string, { count: number; win_rate: number; avg_return: number }>
  daily_results: Array<{
    date: string
    total_value: number
    position_count: number
    position_pct: number
    cash: number
    position_value: number
    return_pct: number
    drawdown: number
    market_rise_ratio: number
  }>
  top_trades: BacktestTrade[]
  worst_trades: BacktestTrade[]
  params?: Record<string, any>
  took_ms?: number
}

// ===== 三买三卖策略 =====

export interface ThreeBuysThreeSellsScanReq {
  min_score?: number
  top_n?: number
  hold_days?: number
  initial_capital?: number
  max_position_pct?: number
  limit?: number
}

export interface ThreeBuysThreeSellsItem {
  code: string
  name: string
  industry: string
  market_cap: number
  close: number
  pct_chg: number
  bias60: number
  ma60: number
  ma60_direction: string
  ma60_slope: number
  stock_type: string
  stock_type_label: string
  signals: Array<{
    type: string
    type_label: string
    trigger_price: number
    position_pct?: number
  }>
  primary_signal_type: string
  primary_signal_label: string
  bottom_pickup: boolean
  macd_divergence: string
  volume_price_divergence: boolean
  score: number
  score_details: string[]
  s1_threshold: number
  stop_price: number
  volume_ratio: number
  atr14: number
  dif: number
  dea: number
  macd_hist: number
  dg_quadrant: string
  dg_available: boolean
  dg_g?: number
  dg_dg?: number
  market_trend: string
  trigger_date: string
}

export interface ThreeBuysThreeSellsScanResp {
  total: number
  items: ThreeBuysThreeSellsItem[]
  took_ms?: number
  scanned_count?: number
  params?: Record<string, any>
  market_trend?: string
}

export interface ThreeBuysThreeSellsBacktestReq extends ThreeBuysThreeSellsScanReq {
  start_date?: string
  end_date?: string
}

export interface ThreeBuysThreeSellsTrade {
  code: string
  name: string
  buy_date: string
  sell_date: string
  buy_price: number
  sell_price: number
  return_pct: number
  score: number
  signal_type: string
  sell_reason: string
  profit: number
}

export interface ThreeBuysThreeSellsBacktestResp {
  total_trades: number
  win_rate: number
  avg_return: number
  avg_win: number
  avg_loss: number
  profit_loss_ratio: number
  max_drawdown: number
  sharpe_ratio: number
  calmar_ratio: number
  annualized_return: number
  max_consecutive_losses: number
  total_fees_est: number
  total_return: number
  final_capital: number
  initial_capital: number
  backtest_days: number
  signal_stats: Record<string, { count: number; win_rate: number; avg_return: number }>
  sell_reason_stats: Record<string, { count: number; win_rate: number; avg_return: number }>
  daily_results: Array<{
    date: string
    total_value: number
    position_count: number
    position_pct: number
    cash: number
    position_value: number
    return_pct: number
    drawdown: number
  }>
  top_trades: ThreeBuysThreeSellsTrade[]
  worst_trades: ThreeBuysThreeSellsTrade[]
  params?: Record<string, any>
  took_ms?: number
}

export const screeningApi = {
  run: (payload: ScreeningRunReq, options?: { timeout?: number }) =>
    ApiClient.post<ScreeningRunResp>('/api/screening/run', payload, { timeout: options?.timeout ?? 120000 }),
  getFields: () => ApiClient.get<FieldConfigResponse>('/api/screening/fields'),
  getIndustries: () => ApiClient.get<IndustriesResponse>('/api/screening/industries'),
  scanLimitUpPullback: (payload: LimitUpPullbackScanReq, options?: { timeout?: number }) =>
    ApiClient.post<LimitUpPullbackScanResp>('/api/screening/limit-up-pullback/scan', payload, { timeout: options?.timeout ?? 180000 }),
  backtestLimitUpPullback: (payload: LimitUpPullbackBacktestReq, options?: { timeout?: number }) =>
    ApiClient.post<LimitUpPullbackBacktestResp>('/api/screening/limit-up-pullback/backtest', payload, { timeout: options?.timeout ?? 300000 }),
  scanThreeBuysThreeSells: (payload: ThreeBuysThreeSellsScanReq, options?: { timeout?: number }) =>
    ApiClient.post<ThreeBuysThreeSellsScanResp>('/api/screening/three-buys-three-sells/scan', payload, { timeout: options?.timeout ?? 180000 }),
  backtestThreeBuysThreeSells: (payload: ThreeBuysThreeSellsBacktestReq, options?: { timeout?: number }) =>
    ApiClient.post<ThreeBuysThreeSellsBacktestResp>('/api/screening/three-buys-three-sells/backtest', payload, { timeout: options?.timeout ?? 600000 })
}

File: app/routers/screening.py
import logging
from fastapi import APIRouter, HTTPException, Depends, Request
from pydantic import BaseModel, Field
from typing import List, Optional, Dict, Any
from app.routers.auth_db import get_current_user

from app.services.enhanced_screening_service import get_enhanced_screening_service
from app.models.screening import (
    ScreeningCondition, ScreeningRequest as NewScreeningRequest,
    ScreeningResponse as NewScreeningResponse, FieldInfo, BASIC_FIELDS_INFO
)

router = APIRouter(tags=["screening"])
logger = logging.getLogger("webapi")

# 筛选字段配置响应模型
class FieldConfigResponse(BaseModel):
    """筛选字段配置"""
    fields: Dict[str, FieldInfo]
    categories: Dict[str, List[str]]

class ScreeningResponse(BaseModel):
    total: int
    items: List[dict]

enhanced_svc = get_enhanced_screening_service()

# 前端字段名 → 数据库字段名 的映射
_FIELD_MAPPING = {
    "market_cap": "total_mv",
    "pe_ratio": "pe",
    "pb_ratio": "pb",
    "turnover": "turnover_rate",
    "change_percent": "pct_chg",
    "price": "close",
    # 直接用数据库字段名也支持（前端直接用）
    "pe": "pe",
    "pb": "pb",
    "total_mv": "total_mv",
    "circ_mv": "circ_mv",
    "turnover_rate": "turnover_rate",
    "pct_chg": "pct_chg",
    "amount": "amount",
    "volume": "volume",
    "close": "close",
    "market": "market",
}

# 前端操作符 → 数据库操作符 的映射
_OP_MAPPING = {
    "eq": "==", "ne": "!=",
    "gte": ">=", "lte": "<=",
    "gt": ">", "lt": "<",
}

# 数据库中可能全部为空的字段（数据尚未填充）——安全跳过
_FLAG_FIELDS = {
    "macd_golden_fork", "ma20_cross", "ma5_cross", "kdj_golden_fork",
    "volume_ratio", "pe_ttm", "pb_mrq",
    "ma20", "rsi14", "kdj_k", "kdj_d", "kdj_j", "dif", "dea", "macd_hist",
}


@router.get("/fields", response_model=FieldConfigResponse)
async def get_screening_fields(user: dict = Depends(get_current_user)):
    """
    获取筛选字段配置
    返回所有可用的筛选字段及其配置信息
    """
    try:
        # 字段分类
        categories = {
            "basic": ["code", "name", "industry", "area", "market"],
            "market_value": ["total_mv", "circ_mv"],
            "financial": ["pe", "pb", "pe_ttm", "pb_mrq", "roe"],
            "trading": ["turnover_rate", "volume_ratio"],
            "price": ["close", "pct_chg", "amount"],
            "technical": ["ma20", "rsi14", "kdj_k", "kdj_d", "kdj_j", "dif", "dea", "macd_hist"]
        }

        return FieldConfigResponse(
            fields=BASIC_FIELDS_INFO,
            categories=categories
        )

    except Exception as e:
        logger.error(f"[get_screening_fields] 获取字段配置失败: {e}", exc_info=True)
        raise HTTPException(status_code=500, detail=str(e))

def _convert_legacy_conditions_to_new_format(legacy_conditions: Dict[str, Any]) -> List[ScreeningCondition]:
    """向后兼容的辅助函数（当前已由 run_screening 直接处理）"""
    conditions = []
    if isinstance(legacy_conditions, dict):
        children = legacy_conditions.get("children", [])
        for child in children:
            if isinstance(child, dict):
                field = child.get("field")
                op = child.get("op")
                value = child.get("value")
                if field and op and value is not None:
                    mapped_field = _FIELD_MAPPING.get(field, field)
                    mapped_op = _OP_MAPPING.get(op, op)
                    conditions.append(ScreeningCondition(
                        field=mapped_field, operator=mapped_op, value=value
                    ))
    return conditions


# 传统筛选接口（保持向后兼容，但使用增强服务）
@router.post("/run", response_model=ScreeningResponse)
async def run_screening(request: Request, user: dict = Depends(get_current_user)):
    try:
        # 直接从原始 JSON body 中提取条件，绕过 Pydantic 模型验证
        raw_body = await request.json()

        # 从 raw_body 中提取条件列表（支持多种格式）
        raw_conditions: list = []
        if "conditions" in raw_body:
            cond = raw_body["conditions"]
            if isinstance(cond, list):
                raw_conditions = cond
            elif isinstance(cond, dict) and "children" in cond:
                raw_conditions = cond.get("children", [])
        elif "children" in raw_body and isinstance(raw_body["children"], list):
            raw_conditions = raw_body["children"]

        # 映射字段名和操作符
        conditions: List[Dict[str, Any]] = []
        for c in raw_conditions:
            if not isinstance(c, dict):
                continue
            op_raw = c.get("op", "==")
            op = _OP_MAPPING.get(str(op_raw).strip().lower(), op_raw)
            fld = _FIELD_MAPPING.get(str(c.get("field", "")), c.get("field", ""))
            value = c.get("value")
            if not fld or value is None:
                continue

            conditions.append({"field": fld, "operator": op, "value": value})

        logger.info(f"[screening] 筛选请求: limit={raw_body.get('limit')}, parsed_conditions={conditions}")

        # 执行筛选
        result = await enhanced_svc.screen_stocks(
            conditions=conditions,
            market=raw_body.get("market", "CN"),
            date=raw_body.get("date"),
            adj=raw_body.get("adj", "qfq"),
            limit=min(int(raw_body.get("limit", 50)), 500),
            offset=max(int(raw_body.get("offset", 0)), 0),
            order_by=[],
            use_database_optimization=True
        )

        logger.info(f"[screening] 完成: total={result.get('total')}, "
                   f"took={result.get('took_ms')}ms, optimization={result.get('optimization_used')}")

        return ScreeningResponse(total=result["total"], items=result["items"])

    except Exception as e:
        logger.error(f"[screening] 处理失败: {e}", exc_info=True)
        raise HTTPException(status_code=500, detail=str(e))


# 新的优化筛选接口
@router.post("/enhanced", response_model=NewScreeningResponse)
async def enhanced_screening(req: NewScreeningRequest, user: dict = Depends(get_current_user)):
    """
    增强的股票筛选接口
    - 支持更丰富的筛选条件格式
    - 自动选择最优的筛选策略（数据库优化 vs 传统方法）
    - 提供详细的性能统计信息
    """
    try:
        logger.info(f"[enhanced_screening] 筛选条件: {len(req.conditions)}个")
        logger.info(f"[enhanced_screening] 排序与分页: order_by={req.order_by}, limit={req.limit}, offset={req.offset}")

        # 执行增强筛选
        result = await enhanced_svc.screen_stocks(
            conditions=req.conditions,
            market=req.market,
            date=req.date,
            adj=req.adj,
            limit=req.limit,
            offset=req.offset,
            order_by=req.order_by,
            use_database_optimization=req.use_database_optimization
        )

        logger.info(f"[enhanced_screening] 筛选完成: total={result.get('total')}, "
                   f"took={result.get('took_ms')}ms, optimization={result.get('optimization_used')}")

        return NewScreeningResponse(
            total=result["total"],
            items=result["items"],
            took_ms=result.get("took_ms"),
            optimization_used=result.get("optimization_used"),
            source=result.get("source")
        )

    except Exception as e:
        logger.error(f"[enhanced_screening] 筛选失败: {e}")
        raise HTTPException(status_code=500, detail=f"增强筛选失败: {str(e)}")


# 获取单个字段的详细信息
@router.get("/fields/{field_name}", response_model=Dict[str, Any])
async def get_field_info(field_name: str, user: dict = Depends(get_current_user)):
    """获取指定字段的详细信息"""
    try:
        field_info = await enhanced_svc.get_field_info(field_name)
        if not field_info:
            raise HTTPException(status_code=404, detail=f"字段 '{field_name}' 不存在")
        return field_info
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"[screening] 获取字段信息失败: {e}")
        raise HTTPException(status_code=500, detail=f"获取字段信息失败: {str(e)}")


# 验证筛选条件
@router.post("/validate", response_model=Dict[str, Any])
async def validate_conditions(conditions: List[ScreeningCondition], user: dict = Depends(get_current_user)):
    """验证筛选条件的有效性"""
    try:
        validation_result = await enhanced_svc.validate_conditions(conditions)
        return validation_result
    except Exception as e:
        logger.error(f"[screening] 验证条件失败: {e}")
        raise HTTPException(status_code=500, detail=f"验证条件失败: {str(e)}")

# 重复定义的旧端点移除（保留带日志的版本）


@router.get("/industries")
async def get_industries(user: dict = Depends(get_current_user)):
    """
    获取数据库中所有可用的行业列表
    从所有数据源中聚合行业分类数据，返回按股票数量排序的行业列表
    """
    try:
        from app.core.database import get_mongo_db

        db = get_mongo_db()
        collection = db["stock_basic_info"]

        # 🔥 从所有有行业数据的数据源中聚合行业信息
        # 先查询有哪些数据源有非空的 industry 字段
        sources_with_data = []
        try:
            source_pipeline = [
                {"$match": {"industry": {"$exists": True, "$nin": [None, "", "未知"]}}},
                {"$group": {"_id": "$source", "count": {"$sum": 1}}},
                {"$sort": {"count": -1}}
            ]
            async for doc in collection.aggregate(source_pipeline):
                sources_with_data.append({"source": doc.get("_id"), "count": doc.get("count", 0)})
        except Exception as e:
            logger.warning(f"[get_industries] 查询数据源分布失败: {e}")

        logger.info(f"[get_industries] 有行业数据的数据源: {sources_with_data}")

        # 聚合查询：按行业分组并统计股票数量（从所有数据源查询）
        pipeline = [
            {
                "$match": {
                    "industry": {"$exists": True, "$nin": [None, "", "未知"]}
                }
            },
            {
                "$group": {
                    "_id": "$industry",
                    "count": {"$sum": 1}
                }
            },
            {"$sort": {"count": -1}},  # 按股票数量降序排序
            {
                "$project": {
                    "industry": "$_id",
                    "count": 1,
                    "_id": 0
                }
            }
        ]

        industries = []
        async for doc in collection.aggregate(pipeline):
            raw_industry = doc.get("industry")
            safe_industry = ""
            try:
                if raw_industry is None:
                    safe_industry = ""
                elif isinstance(raw_industry, float):
                    if raw_industry != raw_industry or raw_industry in (float("inf"), float("-inf")):
                        safe_industry = ""
                    else:
                        safe_industry = str(raw_industry)
                else:
                    safe_industry = str(raw_industry)
            except Exception:
                safe_industry = ""

            raw_count = doc.get("count", 0)
            safe_count = 0
            try:
                if isinstance(raw_count, float):
                    if raw_count != raw_count or raw_count in (float("inf"), float("-inf")):
                        safe_count = 0
                    else:
                        safe_count = int(raw_count)
                else:
                    safe_count = int(raw_count)
            except Exception:
                safe_count = 0

            if safe_industry and safe_count > 0:
                industries.append({
                    "value": safe_industry,
                    "label": safe_industry,
                    "count": safe_count,
                })

        # 确定主要数据源
        primary_source = sources_with_data[0]["source"] if sources_with_data else "unknown"

        logger.info(f"[get_industries] 返回 {len(industries)} 个行业，主要数据源: {primary_source}")

        return {
            "industries": industries,
            "total": len(industries),
            "source": primary_source,
            "sources": sources_with_data
        }

    except Exception as e:
        logger.error(f"[get_industries] 获取行业列表失败: {e}", exc_info=True)
        raise HTTPException(status_code=500, detail=str(e))


# ========== 涨停回调策略 ==========

class LimitUpPullbackRequest(BaseModel):
    """涨停回调策略请求参数（精简版：4个核心参数）"""
    min_score: int = Field(40, ge=0, le=100, description="最低评分阈值")
    top_n: int = Field(10, ge=1, le=50, description="回测时每次最多选股数")
    hold_days: int = Field(20, ge=5, le=30, description="最大持有天数")
    initial_capital: float = Field(1000000, ge=100000, description="初始资金")
    limit: int = Field(50, ge=1, le=200, description="扫描返回数量限制")


class LimitUpPullbackResponse(BaseModel):
    """涨停回调策略响应"""
    total: int = Field(..., description="符合条件的股票总数")
    items: List[dict] = Field(..., description="股票列表")
    took_ms: Optional[int] = Field(None, description="耗时(毫秒)")
    scanned_count: Optional[int] = Field(None, description="扫描的股票总数")
    params: Optional[dict] = Field(None, description="使用的参数")


@router.post("/limit-up-pullback/scan", response_model=LimitUpPullbackResponse)
async def scan_limit_up_pullback(
    req: LimitUpPullbackRequest,
    user: dict = Depends(get_current_user)
):
    """
    涨停回调（龙回头/N字反包）策略选股
    
    策略逻辑：
    1. 筛选最近N天内出现过涨停的股票
    2. 涨停后出现缩量回调（3-5天）
    3. 回调期间出现地量+下影线（左侧买点）
    4. 放量突破5日线（右侧确认买点）
    5. 回调期间不破10日线（生命线）
    """
    try:
        from app.services.limit_up_pullback_service import get_limit_up_pullback_service
        
        service = get_limit_up_pullback_service()
        params = req.model_dump()
        
        logger.info(f"[limit_up_pullback] 扫描请求: {params}")
        
        result = await service.scan_limit_up_pullback(params)
        
        logger.info(f"[limit_up_pullback] 扫描完成: 找到 {result['total']} 只股票, "
                   f"耗时 {result.get('took_ms')}ms")
        
        return LimitUpPullbackResponse(
            total=result["total"],
            items=result["items"],
            took_ms=result.get("took_ms"),
            scanned_count=result.get("scanned_count"),
            params=result.get("params")
        )
        
    except Exception as e:
        logger.error(f"[limit_up_pullback] 扫描失败: {e}", exc_info=True)
        raise HTTPException(status_code=500, detail=f"涨停回调扫描失败: {str(e)}")


class LimitUpPullbackBacktestRequest(LimitUpPullbackRequest):
    """涨停回调策略回测请求参数"""
    start_date: Optional[str] = Field(None, description="回测开始日期 YYYY-MM-DD")
    end_date: Optional[str] = Field(None, description="回测结束日期 YYYY-MM-DD")


class LimitUpPullbackBacktestResponse(BaseModel):
    """涨停回调策略回测响应"""
    total_trades: int = Field(..., description="总交易次数")
    win_rate: float = Field(..., description="胜率(%)")
    avg_return: float = Field(..., description="平均收益(%)")
    avg_win: float = Field(..., description="平均盈利(%)")
    avg_loss: float = Field(..., description="平均亏损(%)")
    profit_loss_ratio: float = Field(0.0, description="盈亏比")
    max_drawdown: float = Field(..., description="最大回撤(%)")
    sharpe_ratio: float = Field(0.0, description="夏普比率")
    calmar_ratio: float = Field(0.0, description="卡玛比率")
    annualized_return: float = Field(0.0, description="年化收益率(%)")
    max_consecutive_losses: int = Field(0, description="最大连续亏损次数")
    total_fees_est: float = Field(0.0, description="估算总手续费")
    total_return: float = Field(..., description="总收益(%)")
    final_capital: float = Field(..., description="最终资金")
    initial_capital: float = Field(..., description="初始资金")
    backtest_days: int = Field(..., description="回测天数")
    signal_stats: dict = Field(default_factory=dict, description="按信号类型统计")
    sell_reason_stats: dict = Field(default_factory=dict, description="按卖出原因统计")
    daily_results: List[dict] = Field(default_factory=list, description="每日结果(前50天)")
    top_trades: List[dict] = Field(default_factory=list, description="盈利最多的20笔")
    worst_trades: List[dict] = Field(default_factory=list, description="亏损最多的20笔")
    params: Optional[dict] = Field(None, description="使用的参数")
    took_ms: Optional[int] = Field(None, description="耗时(毫秒)")


@router.post("/limit-up-pullback/backtest", response_model=LimitUpPullbackBacktestResponse)
async def backtest_limit_up_pullback(
    req: LimitUpPullbackBacktestRequest,
    user: dict = Depends(get_current_user)
):
    """
    涨停回调策略回测
    """
    try:
        from app.services.limit_up_pullback_service import get_limit_up_pullback_service
        
        service = get_limit_up_pullback_service()
        params = req.model_dump()
        
        logger.info(f"[limit_up_pullback_backtest] 回测请求: {params}")
        
        result = await service.backtest(params)
        
        logger.info(f"[limit_up_pullback_backtest] 回测完成: {result['total_trades']} 笔交易, "
                   f"胜率 {result['win_rate']}%, 平均收益 {result['avg_return']}%, "
                   f"耗时 {result.get('took_ms')}ms")
        
        return LimitUpPullbackBacktestResponse(**result)
        
    except Exception as e:
        logger.error(f"[limit_up_pullback_backtest] 回测失败: {e}", exc_info=True)
        raise HTTPException(status_code=500, detail=f"涨停回调回测失败: {str(e)}")


# ========== 三买三卖策略 ==========

class ThreeBuysThreeSellsRequest(BaseModel):
    """三买三卖策略请求参数（精简版：5个核心参数）"""
    min_score: int = Field(50, ge=0, le=100, description="最低信号评分(100分制)")
    top_n: int = Field(10, ge=1, le=50, description="回测时每次最多选股数")
    hold_days: int = Field(60, ge=5, le=120, description="最大持有天数")
    initial_capital: float = Field(1000000, ge=100000, description="初始资金")
    max_position_pct: float = Field(0.15, ge=0.01, le=0.5, description="单股最大仓位比例")
    limit: int = Field(50, ge=1, le=200, description="扫描返回数量限制")


class ThreeBuysThreeSellsResponse(BaseModel):
    """三买三卖策略响应"""
    total: int = Field(..., description="符合条件的股票总数")
    items: List[dict] = Field(..., description="股票列表")
    took_ms: Optional[int] = Field(None, description="耗时(毫秒)")
    scanned_count: Optional[int] = Field(None, description="扫描的股票总数")
    params: Optional[dict] = Field(None, description="使用的参数")
    market_trend: Optional[str] = Field(None, description="大盘趋势")


@router.post("/three-buys-three-sells/scan", response_model=ThreeBuysThreeSellsResponse)
async def scan_three_buys_three_sells(
    req: ThreeBuysThreeSellsRequest,
    user: dict = Depends(get_current_user)
):
    """
    三买三卖交易策略扫描

    三类买点：
    - B1 左侧买点: BIAS(60) ∈ [-30%, -20%]
    - B2 突破买点: 放量 + 中阳 + 站上MA55&MA60
    - B3 回踩买点: 回调至MA60附近 + 放量中阳支撑

    三类卖点：
    - S1 减仓预警: BIAS超阈值 或 GMMA慢组压缩>30%
    - S2 主减仓: 连续2日跌破短期均线组
    - S3 清仓卖出: 跌破MA55&MA60且MA60拐头向下

    安全网: 单日跌幅 > ATR×3 → 强制减仓
    """
    try:
        from app.services.three_buys_three_sells_service import get_three_buys_three_sells_service

        service = get_three_buys_three_sells_service()
        params = req.model_dump()

        logger.info(f"[three_buys_three_sells] 扫描请求: {params}")

        result = await service.scan_three_buys_three_sells(params)

        logger.info(f"[three_buys_three_sells] 扫描完成: 找到 {result['total']} 只股票, "
                   f"耗时 {result.get('took_ms')}ms")

        return ThreeBuysThreeSellsResponse(
            total=result["total"],
            items=result["items"],
            took_ms=result.get("took_ms"),
            scanned_count=result.get("scanned_count"),
            params=result.get("params"),
            market_trend=result.get("market_trend")
        )

    except Exception as e:
        logger.error(f"[three_buys_three_sells] 扫描失败: {e}", exc_info=True)
        raise HTTPException(status_code=500, detail=f"三买三卖扫描失败: {str(e)}")


class ThreeBuysThreeSellsBacktestRequest(ThreeBuysThreeSellsRequest):
    """三买三卖策略回测请求参数"""
    start_date: Optional[str] = Field(None, description="回测开始日期 YYYY-MM-DD")
    end_date: Optional[str] = Field(None, description="回测结束日期 YYYY-MM-DD")


class ThreeBuysThreeSellsBacktestResponse(BaseModel):
    """三买三卖策略回测响应"""
    total_trades: int = Field(..., description="总交易次数")
    win_rate: float = Field(..., description="胜率(%)")
    avg_return: float = Field(..., description="平均收益(%)")
    avg_win: float = Field(..., description="平均盈利(%)")
    avg_loss: float = Field(..., description="平均亏损(%)")
    profit_loss_ratio: float = Field(0.0, description="盈亏比")
    max_drawdown: float = Field(..., description="最大回撤(%)")
    sharpe_ratio: float = Field(0.0, description="夏普比率")
    calmar_ratio: float = Field(0.0, description="卡玛比率")
    annualized_return: float = Field(0.0, description="年化收益率(%)")
    max_consecutive_losses: int = Field(0, description="最大连续亏损次数")
    total_fees_est: float = Field(0.0, description="估算总手续费")
    total_return: float = Field(..., description="总收益(%)")
    final_capital: float = Field(..., description="最终资金")
    initial_capital: float = Field(..., description="初始资金")
    backtest_days: int = Field(..., description="回测天数")
    signal_stats: dict = Field(default_factory=dict, description="按信号类型统计")
    sell_reason_stats: dict = Field(default_factory=dict, description="按卖出原因统计")
    daily_results: List[dict] = Field(default_factory=list, description="每日结果(前50天)")
    top_trades: List[dict] = Field(default_factory=list, description="盈利最多的20笔")
    worst_trades: List[dict] = Field(default_factory=list, description="亏损最多的20笔")
    params: Optional[dict] = Field(None, description="使用的参数")
    took_ms: Optional[int] = Field(None, description="耗时(毫秒)")


@router.post("/three-buys-three-sells/backtest", response_model=ThreeBuysThreeSellsBacktestResponse)
async def backtest_three_buys_three_sells(
    req: ThreeBuysThreeSellsBacktestRequest,
    user: dict = Depends(get_current_user)
):
    """
    三买三卖策略回测
    """
    try:
        from app.services.three_buys_three_sells_service import get_three_buys_three_sells_service

        service = get_three_buys_three_sells_service()
        params = req.model_dump()

        logger.info(f"[three_buys_three_sells_backtest] 回测请求: {params}")

        result = await service.backtest(params)

        logger.info(f"[three_buys_three_sells_backtest] 回测完成: {result['total_trades']} 笔交易, "
                   f"胜率 {result['win_rate']}%, 平均收益 {result['avg_return']}%, "
                   f"耗时 {result.get('took_ms')}ms")

        return ThreeBuysThreeSellsBacktestResponse(**result)

    except Exception as e:
        logger.error(f"[three_buys_three_sells_backtest] 回测失败: {e}", exc_info=True)
        raise HTTPException(status_code=500, detail=f"三买三卖回测失败: {str(e)}")